client/src/visualiser-src/graph-visualiser/data-structure/GraphicalGraph.ts

```diff
@@ -0,0 +1,203 @@
+   * TODO: currently unanimated. Does this need an animation?
+   * @param from
+   * @param to
+   * @returns
+   */
+  addEdge(from: number, to: number): AnimationProducer {
+    this.edges = [...this.edges, { source: from, target: to, weight: 5, isBidirectional: false }];
+
+    // Reload the graph to sync the change in `this.edges` with what's shown
+    // in the visualiser canvas.
+    this.loadGraph();
+    return new GraphAddVertexAnimationProducer();
+  }
+
+  /**
+   * Produces a sequence of animations for visualising DFS.
+   * @remarks
+   * This method wraps around the recursive helper, `dfsRecurse`.
+   * @param src The vertex to start DFS from.
+   * @returns Animation producer with the steps necessary for visualising DFS.
+   */
+  public dfs(src: number): AnimationProducer {
+    const producer = new GraphDfsAnimationProducer();
+    producer.renderDfsCode();
+
+    const visited = new Set<number>();
+    this.dfsRecurse(producer, src, visited);
+
+    return producer;
+  }
+
+  /**
+   * Tells the animation producer to produce the sequence of vertex and edge
+   * highlighting necessary to animate a DFS traversal of the current graph.
+   * @remarks
+   * Assumes that `visited` does not contain `src`.
+   * @param animationProducer
+   * @param src
+   * @param visited A set of all vertices that have been visited through DFS.
+   */
+  private dfsRecurse(
+    animationProducer: GraphDfsAnimationProducer,
+    src: number,
+    visited: Set<number>
+  ): void {
+    // Mark the current vertex as visited.
+    visited.add(src);
+    const vertexElem = this.getDomVertex(src);
+    animationProducer.doAnimationAndHighlight(2, animationProducer.highlightVertex, vertexElem);
+
+    // For each unvisited neighbour, highlight the edge to that neighbour and
+    // launch DFS on them.
+    for (let neighbour = 0; neighbour < this.vertices.length; neighbour += 1) {
+      if (this.isAdjacent(src, neighbour) && !visited.has(neighbour)) {
+        animationProducer.doAnimationAndHighlight(
+          4,
+          animationProducer.highlightEdge,
+          this.getDomEdge(src, neighbour)
+        );
+        this.dfsRecurse(animationProducer, neighbour, visited);
+      }
+    }
+  }
+
+  /**
+   * Determines whether the edge from `v` to `w` exists.
+   * @param v Source vertex.
+   * @param w Destination vertex.
+   * @returns existence of edge v-w.
+   */
+  private isAdjacent(v: number, w: number): boolean {
+    return this.edges.some(
+      (edge) =>
+        (edge.source === v && edge.target === w) ||
+        (edge.source === w && edge.target === v && edge.isBidirectional)
+    );
+  }
+
+  public generate(): void {
+    // TODO: IMPLEMENT ME
+    alert('NOT IMPLEMENTED');
+  }
+
+  public reset(): void {
+    // Reset state.
+    // TODO: IMPLEMENT ME
+    alert('NOT IMPLEMENTED');
+  }
+
+  public get documentation() {
+    return GraphicalGraph.documentation;
+  }
+
+  private getDomVertex(vertex: number): Circle {
```

Comment:
We can remove these methods if we store the svg.js reference upon vertex and edge construction

Reply:
Deleted these methods and added a static method `Vertex.getDomReference` and `Edge.getDomReference` to replace them.